pkg/operator/controller/gatewayclass/servicemeshcontrolplane.go

```diff
@@ -0,0 +1,218 @@
+package gatewayclass
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	gatewayapiv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	maistrav1 "github.com/maistra/istio-operator/pkg/apis/maistra/v1"
+	maistrav2 "github.com/maistra/istio-operator/pkg/apis/maistra/v2"
+
+	"github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// ensureServiceMeshControlPlane attempts to ensure that a
+// servicemeshcontrolplane is present and returns a Boolean indicating whether
+// it exists, the servicemeshcontrolplane if it exists, and an error value.
+func (r *reconciler) ensureServiceMeshControlPlane(ctx context.Context, gatewayclass *gatewayapiv1beta1.GatewayClass) (bool, *maistrav2.ServiceMeshControlPlane, error) {
+	name := controller.ServiceMeshControlPlaneName(r.config.OperatorNamespace, gatewayclass)
+	have, current, err := r.currentServiceMeshControlPlane(ctx, name)
+	if err != nil {
+		return false, nil, err
+	}
+
+	ownerRef := metav1.OwnerReference{
+		APIVersion: gatewayapiv1beta1.SchemeGroupVersion.String(),
+		Kind:       "GatewayClass",
+		Name:       gatewayclass.Name,
+		UID:        gatewayclass.UID,
+	}
+	desired, err := desiredServiceMeshControlPlane(name, ownerRef)
+	if err != nil {
+		return have, current, err
+	}
+
+	switch {
+	case !have:
+		if err := r.createServiceMeshControlPlane(ctx, desired); err != nil {
+			return false, nil, err
+		}
+		return r.currentServiceMeshControlPlane(ctx, name)
+	case have:
+		if updated, err := r.updateServiceMeshControlPlane(ctx, current, desired); err != nil {
+			return have, current, err
+		} else if updated {
+			return r.currentServiceMeshControlPlane(ctx, name)
+		}
+	}
+	return true, current, nil
+}
+
+// desiredServiceMeshControlPlane returns the desired servicemeshcontrolplane.
+func desiredServiceMeshControlPlane(name types.NamespacedName, ownerRef metav1.OwnerReference) (*maistrav2.ServiceMeshControlPlane, error) {
+	pilotContainerEnv := map[string]string{
+		"PILOT_ENABLE_GATEWAY_API":                       "true",
+		"PILOT_ENABLE_GATEWAY_API_DEPLOYMENT_CONTROLLER": "true",
+		"PILOT_ENABLE_GATEWAY_API_STATUS":                "true",
+		"PILOT_GATEWAY_API_CONTROLLER_NAME":              OpenShiftGatewayClassControllerName,
+	}
+	f := false
+	t := true
+	smcp := maistrav2.ServiceMeshControlPlane{
```

Comment:
It seems risky to hard code the SMCP values before we have attained release alignment.  Things could change on OSSM between OpenShift releases.  Is there any way to do this without hard-coding it ?

Reply:
What do you have in mind?  Do you envision that the ingress operator could download the SMCP configuration from some out-of-payload source?  That seems like it might be fragile and a potential attack vector, so I'd want us to design any such mechanism very carefully.  For now, I'd rather not deviate to far from the design in the EP.  If something changes in the SMCP definition, we can address it as a bug fix.